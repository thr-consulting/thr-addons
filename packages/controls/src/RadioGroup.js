// @flow

import React, {Component, Children} from 'react';
import type {ChildrenArray} from 'react';
import debug from 'debug';
import {Form} from 'semantic-ui-react';

const d = debug('thx:controls:RadioGroup');

type Props = {
	children: ChildrenArray<*>,
	onChange: (value: boolean | number | string | Object) => {},
	value: | boolean | number | string | Object,
};

/**
 * Groups React Semantic UI Radio elements into a single group
 * @class
 * @property {Component[]} children - Radio or Form.Radio components
 * @property {onChange} onChange - Standard onChange handler
 * @property {bool|number|string|Object} value - The currently selected radio item
 */
export default class RadioGroup extends Component<Props> {
	static defaultProps = {
		children: null,
		onChange: null,
		value: null,
	};

	props: Props;

	handleChange = (e: Event, {value}: {value: any}) => {
		d('Value changed to:', value);
		if (this.props.onChange) this.props.onChange(value);
	}

	render() {
		const {children, value, onChange, ...rest} = this.props;

		return (
			<Form.Group {...rest}>
				{Children.map(children, child => {
					if (child.type.name === 'FormRadio' || child.type.name === 'Radio') {
						return React.cloneElement(child, {
							onChange: this.handleChange,
							checked: value === child.props.value,
						});
					}
					return child;
				})}
			</Form.Group>
		);
	}
}
