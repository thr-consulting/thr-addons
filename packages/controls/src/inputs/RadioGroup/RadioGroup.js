import React, {Component, Children} from 'react';
import {Form, Radio, FormRadio} from 'semantic-ui-react';

// type Props = {
// 	/** Any components but handles Radio components specially. */
// 	children?: ChildrenArray<*>,
// 	/** Called when one of the Radio boxes is selected. */
// 	onChange?: (value: boolean | number | string | Object) => void,
// 	onBlur?: () => void,
// 	/** This value will select the correct Radio component with the matching value */
// 	value?: | boolean | number | string | Object,
// };

/**
 * Groups React Semantic UI Radio elements into a single group
 * @class
 * @property {Component[]} children - Radio or Form.Radio components
 * @property {onChange} onChange - Standard onChange handler
 * @property {bool|number|string|Object} value - The currently selected radio item
 */
export default class RadioGroup extends Component {
	static displayName = 'RadioGroup';

	static defaultProps = {
		children: null,
		onChange: null,
		onBlur: null,
		value: null,
	};

	handleChange = (e, {value}) => {
		if (this.props.onChange) this.props.onChange(value);
	};

	render() {
		const {children, value, onChange, onBlur, ...rest} = this.props;

		return (
			<Form.Group {...rest} onBlur={onBlur}>
				{Children.map(children, child => {
					if (child.type === FormRadio || child.type === Radio || child.type === Form.Radio) {
						return React.cloneElement(child, {
							onChange: this.handleChange,
							checked: value === child.props.value,
						});
					}
					return child;
				})}
			</Form.Group>
		);
	}
}
