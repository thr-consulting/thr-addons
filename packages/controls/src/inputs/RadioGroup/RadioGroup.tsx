import debug from 'debug';
import React, {Children, cloneElement} from 'react';
import {Form, Radio, FormRadio, FormGroupProps} from 'semantic-ui-react';

const d = debug('thx.controls.inputs.RadioGroup');

export interface RadioGroupProps {
	children?: JSX.Element[] | JSX.Element;
	onChange?: (value?: string | number) => void;
	value?: string | number;
}

export function RadioGroup(props: RadioGroupProps & Omit<FormGroupProps, 'children'>) {
	const {children, onChange, value, ...rest} = props;

	return (
		<Form.Group {...rest}>
			{Children.map(props.children, child => {
				if (child?.type === FormRadio || child?.type === Radio || child?.type === Form.Radio) {
					return cloneElement(child, {
						onChange: (ev: any, args: {value?: string | number}) => {
							if (onChange) onChange(args.value);
						},
						checked: value === child.props.value,
					});
				}
				return child;
			})}
		</Form.Group>
	);
}
