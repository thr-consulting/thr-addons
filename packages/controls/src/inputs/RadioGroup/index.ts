export {RadioGroupProps, RadioGroup} from './RadioGroup';
