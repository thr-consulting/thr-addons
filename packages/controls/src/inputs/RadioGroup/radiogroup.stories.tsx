import {Radio, RadioGroup} from '@mantine/core';
import {useArgs} from '@storybook/client-api';
import type {ComponentStory, Meta} from '@storybook/react';
import debug from 'debug';
import {storyDecorator} from '../../storyDecorator';

const d = debug('thx.controls.inputs.RadioGroup.radiogroup.stories');

export default {
	title: 'Inputs/RadioGroup',
	argTypes: {
		value: {type: 'string'},
		onChange: {type: 'function'},
	},
	decorators: [storyDecorator],
} as Meta;

const t: ComponentStory<typeof RadioGroup> = args => {
	// eslint-disable-next-line react-hooks/rules-of-hooks
	const [, updateArgs] = useArgs();

	return (
		<RadioGroup
			{...args}
			onChange={value => {
				d(value);
				updateArgs({value});
			}}
		>
			<Radio value="a" label="Choice A" />
			<Radio value="b" label="Choice B" />
		</RadioGroup>
	);
};

export const Main = t.bind({});
Main.args = {
	value: undefined,
};
