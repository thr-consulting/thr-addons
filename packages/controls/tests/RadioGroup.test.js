import React from 'react';
import {shallow, mount} from 'enzyme';
import {Form} from 'semantic-ui-react';
import RadioGroup from '../src/RadioGroup';

describe('RadioGroup', () => {
	it('should render without throwing an error', () => {
		expect(shallow(
			<RadioGroup>
				<Form.Radio label="ABC" value="abc"/>
				<Form.Radio label="DEF" value="def"/>
			</RadioGroup>)).toMatchSnapshot();
	});

	it('should respond to changes', () => {
		const mockOnChange = jest.fn();

		const wrapper = mount(
			<RadioGroup onChange={mockOnChange}>
				<Form.Radio label="ABC" value="abc"/>
				<Form.Radio label="DEF" value="def"/>
			</RadioGroup>,
		);
		const u = wrapper.find('.ui.radio.checkbox').first();
		u.simulate('click');

		expect(mockOnChange.mock.calls.length).toBe(1);
		expect(mockOnChange.mock.calls[0][0]).toEqual('abc');
	});
});
