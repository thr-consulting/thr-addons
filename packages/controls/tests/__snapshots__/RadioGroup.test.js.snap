// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should render without throwing an error 1`] = `
<FormGroup>
  <FormRadio
    as={[Function]}
    checked={false}
    control={[Function]}
    label="ABC"
    onChange={[Function]}
    value="abc"
  />
  <FormRadio
    as={[Function]}
    checked={false}
    control={[Function]}
    label="DEF"
    onChange={[Function]}
    value="def"
  />
</FormGroup>
`;
